src/modm/platform/adc/stm32f0/adc_impl.hpp.in

```diff
@@ -360,3 +367,127 @@ modm::platform::Adc{{ id }}::setWaitMode(const bool enable)
 		ADC1->CFGR1 &= ~ADC_CFGR1_WAIT;
 	}
 }
+
+%% if target.family in ["g0"]
+uintptr_t
+modm::platform::Adc{{ id }}::getDataRegisterAddress()
+{
+	return (uintptr_t) &(ADC1->DR);
+}
+
+void
+modm::platform::Adc{{ id }}::enableRegularConversionExternalTrigger(
+	ExternalTriggerPolarity externalTriggerPolarity,
+	RegularConversionExternalTrigger regularConversionExternalTrigger)
+{
+	const auto polarity =
+		(static_cast<uint32_t>(externalTriggerPolarity) << ADC_CFGR1_EXTEN_Pos);
+	const auto externalTrigger =
+		(static_cast<uint32_t>(regularConversionExternalTrigger) << ADC_CFGR1_EXTSEL_Pos);
+	const auto mask = ADC_CFGR1_EXTEN_Msk | ADC_CFGR1_EXTSEL_Msk;
+	ADC1->CFGR1 = (ADC1->CFGR1 & ~mask) | polarity | externalTrigger;
+}
+
+void
+modm::platform::Adc{{ id }}::enableDmaMode()
+{
+	ADC1->CFGR1 |= ADC_CFGR1_DMAEN;
+}
+
+void
+modm::platform::Adc{{ id }}::disableDmaMode()
+{
+	ADC1->CFGR1 &= ~ADC_CFGR1_DMAEN;
+}
+
+bool
+modm::platform::Adc{{ id }}::getAdcEnabled()
+{
+	return (ADC1->CR & ADC_CR_ADEN);
+}
+
+
+void
+modm::platform::Adc{{ id }}::enableDmaRequests()
+{
+	ADC1->CFGR1 |= ADC_CFGR1_DMACFG;
+}
+
+void
+modm::platform::Adc{{ id }}::disableDmaRequests()
+{
+	ADC1->CFGR1 &= ~ADC_CFGR1_DMACFG;
+}
+
+void
+modm::platform::Adc{{ id }}::waitChannelConfigReady()
+{
+	while (!(ADC1->ISR & ADC_ISR_CCRDY));
+}
+
+void
+modm::platform::Adc{{ id }}::resetChannelConfigReady(){ADC1->ISR &= ~ADC_ISR_CCRDY;}
+
+void
+modm::platform::Adc{{ id }}::enableScanMode()
+{
+	resetChannelConfigReady();
+	ADC1->CFGR1 |= ADC_CFGR1_CHSELRMOD;
+	waitChannelConfigReady();
+
+	resetChannelConfigReady();
+	ADC1->CHSELR = uint32_t(0xFFFFFFFF);
+	waitChannelConfigReady();
+}
+
+void
+modm::platform::Adc{{ id }}::disableScanMode()
+{
+	ADC1->CFGR1 &= ~ADC_CFGR1_CHSELRMOD;
+}
```

Comment:
I'm not sure the sequence API taken from the F4 ADC driver really fits the F0/G0 peripheral that well.

The reference manual states there are two different scan modes available on G0 devices, one where `CHSELRMOD` is set and one where it is not. The one implemented here with `CHSELRMOD` allows to sample up to 8 channels in arbitrary order, but ADC channels 15-18 can't be read and it's not supported on F0 devices.
The other method will always sample channels in order but also works with all inputs  and on all devices this driver supports.

For a user looking at the peripheral registers the API can easily be confusing because `CHSELRMOD` doesn't control if sequential scanning of channels is enabled but which method is used. A function `disableScanMode()` that clears `CHSELRMOD` doesn't really make sense from a hardware point of view.

I'd recommend to adapt the API in a way that it maps to the hardware functionality more directly and makes it harder to misuse. Furthermore we could make all inputs up to 18 accessible in scan mode and support F0 devices as well. All the various STM32 ADC peripherals are different enough that having one fully consistent API won't really work out if it should provide more advanced features.

I haven't worked out all details, but we could have `setChannel` configure the ADC in single channel mode, one `setChannelSequence` that takes channel bit flags for the non-`CHSELRMOD` method and `setChannelSequence` taking a `std::span<Channel>` for the method in this PR. The latter wouldn't be available on F0. `enableScanMode()` and `disableScanMode()` wouldn't be required anymore, you'd get scan mode if you configure a sequence and a single channel conversion otherwise. 

Probably `waitChannelConfigReady()` and `resetChannelConfigReady()` should be handled internally inside the driver because they aren't required on F0 devices at all. I'm not sure if there is a reason to interact with them directly as a user.

What do you think?